File: leetcode/0316-remove-duplicate-letters.ts
function removeDuplicateLetters(s: string): string {
    const counter: number[] = new Array(26).fill(0);

    for (const c of s) {
        counter[c.charCodeAt(0) - 97]++;
    }

    const stack: string[] = [];
    const visited: boolean[] = new Array(26).fill(false);

    for (const c of s) {
        const index = c.charCodeAt(0) - 97;
        counter[index]--;

        if (visited[index]) {
            continue;
        }

        while (
            stack.length > 0 &&
            c < stack[stack.length - 1] &&
            counter[
                stack[stack.length - 1].charCodeAt(0) - 97
            ] > 0
        ) {
            const top = stack.pop()!;
            visited[top.charCodeAt(0) - 97] = false;
        }

        stack.push(c);
        visited[index] = true;
    }

    return stack.join("");
}

File: 0316-remove-duplicate-letters/0316-remove-duplicate-letters.java
import java.util.*;

class Solution {
    public String removeDuplicateLetters(String s) {
        int[] counter = new int[26];

        for (char c : s.toCharArray()) {
            counter[c - 'a']++;
        }

        Stack<Character> stack = new Stack<>();
        boolean[] visited = new boolean[26];

        for (char c : s.toCharArray()) {
            counter[c - 'a']--;

            if (visited[c - 'a']) {
                continue;
            }

            while (!stack.isEmpty()
                    && c < stack.peek()
                    && counter[stack.peek() - 'a'] > 0) {

                char top = stack.pop();
                visited[top - 'a'] = false;
            }

            stack.push(c);
            visited[c - 'a'] = true;
        }

        StringBuilder ans = new StringBuilder();

        while (!stack.isEmpty()) {
            ans.append(stack.pop());
        }

        return ans.reverse().toString();
    }
}